File: server-azure/src/model/game/user-game.ts
import type { Board } from './board';
import { CatanError } from '../../core/catan-error';
import type { Color } from './color';
import type { Player } from './player';
import type { Writable } from 'ts-essentials';

export type UserGame = Readonly<{
  id: string;
  game: string;
  user: string;
  color: Color;
  myTurn: boolean;
  completed: boolean;
}>;

export function createUserGame(board: Board, color: Color): UserGame {
  return createUserGameInternal(board, getPlayer(board, color), false);
}

function createUserGameInternal(board: Board, player: Player, completed: boolean): UserGame {
  return {
    id: generateUserGameId(board.id, player.id),
    game: board.id,
    user: player.id,
    color: player.color,
    myTurn: false,
    completed,
  };
}

export function generateUserGameId(game: string, user: string): string {
  return `${game}:${user}`;
}

export function createUserGamesFromBoard(board: Board): UserGame[] {
  const completed = board.completed !== 0;
  const ugs = board.players.map(player => createUserGameInternal(board, player, completed));
  if (!completed) {
    (ugs[0] as Writable<UserGame>).myTurn = true;
  }
  return ugs;
}

function getPlayer(board: Board, color: Color): Player {
  const player = board.players.find(player => player.color === color);
  if (!player) {
    throw new CatanError('Got null color for UserGame');
  }
  return player;
}

File: server/src/main/java/org/codi/catan/model/game/UserGame.java
/*
 * @author the-codinator
 * created on 2020/8/8
 */

package org.codi.catan.model.game;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.codi.catan.core.CatanException;
import org.codi.catan.model.core.IdentifiableEntity;

@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
public class UserGame implements IdentifiableEntity {

    private String id;
    private String game;
    private String user;
    private Color color;
    private boolean myTurn;
    private boolean completed;

    public UserGame(Board board, Color color) throws CatanException {
        this(board, getPlayer(board, color), false);
    }

    private UserGame(Board board, Player player, boolean completed) {
        this(generateId(board.getId(), player.getId()), board.getId(), player.getId(), player.getColor(), false,
            completed);
    }

    public static String generateId(String game, String user) {
        return game + ':' + user;
    }

    /**
     * Create UserGame object for each player in the game
     */
    public static UserGame[] createUserGamesFromBoard(Board board) {
        boolean completed = board.getCompleted() != 0;
        UserGame[] ugs = new UserGame[board.getPlayers().length];
        for (int i = 0; i < ugs.length; i++) {
            ugs[i] = new UserGame(board, board.getPlayers()[i], completed);
        }
        if (!completed) {
            ugs[0].setMyTurn(true);
        }
        return ugs;
    }

    private static Player getPlayer(Board board, Color color) throws CatanException {
        Player[] players = board.getPlayers();
        for (Player player : players) {
            if (player.getColor() == color) {
                return player;
            }
        }
        throw new CatanException("Got null color for UserGame");
    }
}
